plugins/logitech-bulkcontroller/fu-logitech-bulkcontroller-device.c

```diff
@@ -734,3 +745,198 @@ fu_logitech_bulkcontroller_device_class_init(FuLogitechBulkcontrollerDeviceClass
+			if (!fu_logitech_bulkcontroller_device_send_sync_cmd(self,
+									     CMD_ACK,
+									     buf_ack,
+									     &error))
+				return error;
+			break;
+		case CMD_UNINIT_BUFFER:
+			fu_byte_array_append_uint32(buf_ack, cmd_tmp, G_LITTLE_ENDIAN);
+			if (!fu_logitech_bulkcontroller_device_send_sync_cmd(self,
+									     CMD_ACK,
+									     buf_ack,
+									     &error))
+				return error;
+			g_cond_signal(&self->sync_transfer_cond);
+			break;
+		default:
+			break;
+		}
+	}
+	return NULL;
+}
+
+static gboolean
+fu_logitech_bulkcontroller_device_get_data(FuDevice *device, GError **error)
+{
+	FuLogitechBulkcontrollerDevice *self = FU_LOGITECH_BULKCONTROLLER_DEVICE(device);
+	g_autoptr(GByteArray) device_request = g_byte_array_new();
+	g_autoptr(GByteArray) decoded_pkt = g_byte_array_new();
+	FuLogitechBulkcontrollerProtoId proto_id = kProtoId_UnknownId;
+
+	/* sending GetDeviceInfoRequest */
+	device_request = proto_manager_generate_get_device_info_request();
+	if (!fu_logitech_bulkcontroller_device_send_sync_cmd(self,
+							     CMD_BUFFER_WRITE,
+							     device_request,
+							     error)) {
+		g_prefix_error(error,
+			       "error in sending buffer write packet for device info request: ");
+		return FALSE;
+	}
+	if (!g_mutex_trylock(&self->sync_transfer_mutex)) {
+		g_prefix_error(error, "error in acquiring lock for device info request: ");
+		return FALSE;
+	}
+	if (!g_cond_wait_until(&self->sync_transfer_cond,
+			       &self->sync_transfer_mutex,
+			       g_get_monotonic_time() + THREE_SECONDS)) {
+		g_mutex_unlock(&self->sync_transfer_mutex);
+		g_prefix_error(
+		    error,
+		    "timing out, failed to receive response from device for device info request: ");
+		return FALSE;
+	}
+	g_mutex_unlock(&self->sync_transfer_mutex);
+	decoded_pkt = proto_manager_decode_message(self->device_response->data,
+						   self->device_response->len,
+						   &proto_id,
+						   error);
+	g_byte_array_free(self->device_response, TRUE);
+	if (decoded_pkt == NULL) {
+		g_prefix_error(error, "error in unpacking response for device info request: ");
+		return FALSE;
+	}
+	g_debug("Received device response. data: %s, length: %u",
+		(const gchar *)decoded_pkt->data,
+		decoded_pkt->len);
+
+	if (proto_id != kProtoId_GetDeviceInfoResponse) {
+		g_set_error_literal(error,
+				    G_IO_ERROR,
+				    G_IO_ERROR_INVALID_DATA,
+				    "incorrect response for device info request: ");
+		return FALSE;
+	}
+	if (!fu_logitech_bulkcontroller_device_json_parser(device, decoded_pkt, error))
+		return FALSE;
+
+	/* success */
+	return TRUE;
+}
+
+static void
+fu_logitech_bulkcontroller_device_startlistening_sync(FuDevice *device)
+{
+	FuLogitechBulkcontrollerDevice *self = FU_LOGITECH_BULKCONTROLLER_DEVICE(device);
+
+	self->is_sync_transfer_in_progress = TRUE;
+	self->sync_thread = g_thread_new("Sync Transfer thread",
+					 &fu_logitech_bulkcontroller_device_sync_thread,
+					 device);
```

Comment:
I still don't see why you need a thread. We've got multiple `GMainLoop`s and multiple `GMainContext`s, and I don't see how it makes anything possible that you can't do before.

Reply:
Will explore GMainLoops and multiple GMainContexts. Any additional pointers/examples for these objects?

Reply:
Exploring it. We have some success with GMainLoop, but still seeing some issues. Will reach out to you as needed.